return internal error code, if failed

RonEld opened the pull request:
## Description
in case `mbedtls_aes_crypt_ecb` failed, return an error from other modes
calling it
Resolves #1092 

## Status
**READY**

## Requires Backporting
Yes 
Which branch?
mbedtls-2.1
mbedtls-1.3

## Migrations
NO


## Todos
- [ ] Tests
- [ ] Documentation
- [x] Changelog updated
- [ ] Backported

hanno-arm commented on ChangeLog:
```diff
@@ -1,5 +1,11 @@
+
+Bugfix
+   * return error code of `mbedtls_aes_crypt_ecb` if failed, when calling it from
+     other modes. FIx for #1092 reported by adustm
```
Two typos here

hanno-arm commented on library/aes.c:
```diff
@@ -979,6 +995,9 @@ int mbedtls_aes_crypt_cfb128( mbedtls_aes_context *ctx,
+end:
+    if ( ret != 0 )
+        mbedtls_zeroize( output, length );
     return( 0 );
```
This should be `return( ret );`

hanno-arm commented on library/aes.c:
```diff
@@ -1010,7 +1032,9 @@ int mbedtls_aes_crypt_cfb8( mbedtls_aes_context *ctx,
+end:
+    if ( ret != 0 )
+        mbedtls_zeroize( output, length );
     return( 0 );
```
This should be `return( ret );`

hanno-arm commented on library/aes.c:
```diff
@@ -1046,7 +1072,9 @@ int mbedtls_aes_crypt_ctr( mbedtls_aes_context *ctx,
+end:
+    if ( ret != 0 )
+        mbedtls_zeroize( output, length );
     return( 0 );
```
This should be `return( ret );`

andresag01 commented on library/aes.c:
```diff
@@ -1027,13 +1051,15 @@ int mbedtls_aes_crypt_ctr( mbedtls_aes_context *ctx,
     {
         if( n == 0 ) {
-            mbedtls_aes_crypt_ecb( ctx, MBEDTLS_AES_ENCRYPT, nonce_counter, stream_block );
+            ret = mbedtls_aes_crypt_ecb( ctx, MBEDTLS_AES_ENCRYPT, nonce_counter, stream_block );
```
Style: This line is a bit too long. Could you please break it up?

hanno-arm commented on library/aes.c:
```diff
@@ -880,7 +880,7 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
                     unsigned char *output )
 {
-    int i;
+    int i, ret;
```
`ret` should be initialized as the body of the function doesn't alter it for `length == 0`.

hanno-arm commented on library/aes.c:
```diff
@@ -947,15 +955,19 @@ int mbedtls_aes_crypt_cfb128( mbedtls_aes_context *ctx,
                        unsigned char *output )
 {
-    int c;
+    int c, ret;
```
`ret` should be initialized here, too.

hanno-arm commented on library/aes.c:
```diff
@@ -994,11 +1013,14 @@ int mbedtls_aes_crypt_cfb8( mbedtls_aes_context *ctx,
 {
     unsigned char c;
     unsigned char ov[17];
+    int ret;
```
`ret` should be initialized here, too.

hanno-arm commented on library/aes.c:
```diff
@@ -1027,13 +1051,15 @@ int mbedtls_aes_crypt_ctr( mbedtls_aes_context *ctx,
                        unsigned char *output )
 {
-    int c, i;
+    int c, i, ret;
```
`ret` should be initialized here, too.

RonEld commented:
Hi @andresag01 @hanno-arm Thank you for your comments. I agree we should change other files as well

RonEld commented:
Hi @andresag01 @hanno-arm I modified according to your comments
Note that there are still locations where we don't check ther return code:
~~~
> grep -rIn "mbedtls_aes_crypt_\(ecb\|cbc\|ctr\)" ./library/*
./library/aes.c:1308:                mbedtls_aes_crypt_ecb( &ctx, v, buf, buf );
./library/aes.c:1324:                mbedtls_aes_crypt_ecb( &ctx, v, buf, buf );
./library/aes.c:1365:                mbedtls_aes_crypt_cbc( &ctx, v, 16, iv, buf, buf );
./library/aes.c:1384:                mbedtls_aes_crypt_cbc( &ctx, v, 16, iv, buf, buf );
./library/aes.c:1489:            mbedtls_aes_crypt_ctr( &ctx, len, &offset, nonce_counter, stream_block,
./library/aes.c:1506:            mbedtls_aes_crypt_ctr( &ctx, len, &offset, nonce_counter, stream_block,
./library/pem.c:195:    mbedtls_aes_crypt_cbc( &aes_ctx, MBEDTLS_AES_DECRYPT, buflen,
~~~
In `aes.c` these are in the selftest, which is already addressed in #964 and I prefer to avoid the conflicts
In `pem.c` , the call is in `static void pem_aes_decrypt()` , which we may consider changing its return value as well, but I think this should be done in a different PR

andresag01 commented on ChangeLog:
```diff
@@ -3,8 +3,8 @@ mbed TLS ChangeLog (Sorted per branch, date)
 Bugfix
-   * return error code of `mbedtls_aes_crypt_ecb` if failed, when calling it from
-     other modes. FIx for #1092 reported by adustm
+   * Return error code of `mbedtls_aes_crypt_ecb` if failed, when calling it from
```
Sorry for not pointing it out later, but please note that ChangeLog is not markdown and the \` character has no efect there. I would suggest ` ``mbedtls_aes_crypt_ecb`` ` -> `mbedtls_aes_crypt_ecb()`

RonEld commented:
I fixed the ChangeLog
@andresag01 @hanno-arm please approve

andresag01 commented on library/ctr_drbg.c:
```diff
@@ -180,8 +182,9 @@ static int block_cipher_df( unsigned char *output,
-    mbedtls_aes_setkey_enc( &aes_ctx, key, MBEDTLS_CTR_DRBG_KEYBITS );
-
+    ret = mbedtls_aes_setkey_enc( &aes_ctx, key, MBEDTLS_CTR_DRBG_KEYBITS );
+    if ( ret != 0 )
```
Style: Please remove ` ` between `if` and `(`

andresag01 commented on library/ctr_drbg.c:
```diff
@@ -261,10 +273,13 @@ static int ctr_drbg_update_internal( mbedtls_ctr_drbg_context *ctx,
+    return( ret );
 }
 
 void mbedtls_ctr_drbg_update( mbedtls_ctr_drbg_context *ctx,
```
I know that its slightly pointless to check the return codes of block_cipher_df() and ctr_drbg_update_internal() as mbedtls_ctr_drbg_update() does not have a return value. However, it would be nice if we could at least abort the update if either of the calls fails. Also, it would be good to add a comment nearby explaning the situation (and perhaps raise a ticket as the API might need to be modified).

andresag01 replied:
It would be good to check the return values of of the functions ctr_drbg_update_internal() and ctr_drbg_update_internal() called from `mbedtls_ctr_drbg_reseed()` as well.

andresag01 replied:
@RonEld: I dont think this was fully addressed. There are still a couple of unchecked return values [here](https://github.com/ARMmbed/mbedtls/pull/1096/commits/d66b1b51207e05575ab939901e6c12fe8c827f4e#diff-af7ff720e446d17cba06099cb0742ec4R347) and [here](https://github.com/ARMmbed/mbedtls/pull/1096/commits/d66b1b51207e05575ab939901e6c12fe8c827f4e#diff-af7ff720e446d17cba06099cb0742ec4R352). Thanks!

RonEld replied:
sorry, I missed this

andresag01 commented:
@RonEld: Now that I recall, the return value for the PEM functions has been fixed in a separate PR. Please refer to https://github.com/ARMmbed/mbedtls/pull/828

RonEld commented:
@andresag01 I addressed your comments. Please approve

andresag01 commented on library/ctr_drbg.c:
```diff
@@ -294,8 +295,17 @@ void mbedtls_ctr_drbg_update( mbedtls_ctr_drbg_context *ctx,
+        if( ret != 0)
+            return;
+        ret = ctr_drbg_update_internal( ctx, add_input );
+        if( ret != 0)
```
Style: This might not even be worth pointing out, but there is a missing ` ` before `)`...

RonEld commented:
@andresag01 Thanks for your comments. I missed that part.
Please review

hanno-arm commented on library/ctr_drbg.c:
```diff
@@ -319,12 +345,15 @@ int mbedtls_ctr_drbg_reseed( mbedtls_ctr_drbg_context *ctx,
-    ctr_drbg_update_internal( ctx, seed );
+    ret = ctr_drbg_update_internal( ctx, seed );
+    if( ret != 0 )
+        return ret;
```
Minor: brackets missing around `ret` here and a few lines above.

RonEld commented:
I have added brackets as your comment.
Please approve

hanno-arm commented on library/aes.c:
```diff
@@ -979,7 +995,10 @@ int mbedtls_aes_crypt_cfb128( mbedtls_aes_context *ctx,
-    return( 0 );
+end:
+    if ( ret != 0 )
+        mbedtls_zeroize( output, length );
```
Sorry for not bringing this up earlier, but this probably does not do what it is intended to do: The `output` and `length` values are updated during the operations, so this zeroization will clear the _remaining_, so far unwritten, output buffer, while the part that has already been written is left untouched. Also, the loops decrease `length` at the beginning of the iteration while `output` is increased at the end, so the zeroization will omit one byte at the end.

RonEld replied:
Good catch! I missed that for some reason. I'll fix that.

RonEld commented:
Hi @hanno-arm I fixed according to your comment
Please review

sbutcher-arm commented:
@RonEld - Could you please rebase to remove the conflicts? We'll then need to arrange approvals again.

RonEld commented:
@sbutcher-arm I have rebased and fixed errors after rebase

@andresag01 @hanno-arm Please review

andresag01 commented on library/ctr_drbg.c:
```diff
@@ -90,15 +90,13 @@ int mbedtls_ctr_drbg_seed_entropy_len(
-    }
+    if( ( ret = mbedtls_aes_setkey_enc( &ctx->aes_ctx,
+                                        key, MBEDTLS_CTR_DRBG_KEYBITS ) ) != 0 )
+            return( ret );
```
Style: Could you please reduce the indentation here by 1 level?

RonEld replied:
done

andresag01 commented on library/ctr_drbg.c:
```diff
@@ -313,8 +312,17 @@ void mbedtls_ctr_drbg_update( mbedtls_ctr_drbg_context *ctx,
+            return;
+        ret = ctr_drbg_update_internal( ctx, add_input );
+        if( ret != 0 )
+            return;
```
Do we have a bug report about the missing return code in the function signature?

RonEld replied:
We do now, in #1798

RonEld commented:
@andresag01 I have addressed your comments, please review

sbutcher-arm commented on library/ctr_drbg.c:
```diff
@@ -90,15 +90,13 @@ int mbedtls_ctr_drbg_seed_entropy_len(
-    }
 
     if( ( ret = mbedtls_ctr_drbg_reseed( ctx, custom, len ) ) != 0 )
-    {
```
Why are we removing brackets here? I know we don't often have them for single statements, but a lot of coding standards require them as defensive programming. Can you put them back in please?

sbutcher-arm commented on library/ctr_drbg.c:
```diff
@@ -313,8 +312,17 @@ void mbedtls_ctr_drbg_update( mbedtls_ctr_drbg_context *ctx,
-        ctr_drbg_update_internal( ctx, add_input );
+        /*
+         * block_cipher_df() and ctr_drbg_update_internal() may return error codes.
+         * If they return error, abort the operation, since this public API
```
Grammar - "If they return *an* error, abort the operation, since this *function* does not have a return code, we can't return *an* error code to the caller."

sbutcher-arm commented on ChangeLog:
```diff
@@ -1,5 +1,11 @@
+= mbed TLS x.x.x branch released xxxx-xx-xx
+
+Bugfix
+   * Return error code of mbedtls_aes_crypt_ecb() if failed, when calling it from
```
Grammar - "Handle the error if mbedtls_aes_crypt_ecb() fails, when calling it from other modes and ciphers. Fix for #1092 reported by adustm"

RonEld commented:
@sbutcher-arm I addressed your comments

cc @andresag01

RonEld commented:
I have rebased to add all the new features that enhanced AES, such as OFB

RonEld commented:
After rebase, there isn't any new calls to `mbedtls_aes_crypt_ecb()` without checking return code.

cc @Patater

Patater commented on library/aes.c:
```diff
@@ -1014,8 +1014,10 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+    int i, ret = 0;
     unsigned char temp[16];
+    const size_t orig_length = length;
+    unsigned char* out = output;
```
Style: Use `char *out` instead of `char* out` (here and in other places).

RonEld commented:
@Patater I addressed your comment

andresag01 commented on library/aes.c:
```diff
@@ -1065,7 +1072,10 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
 
-    return( 0 );
+end:
+    if ( ret != 0 )
```
Style: Please remove space between `if` and `(`

andresag01 commented on library/aes.c:
```diff
@@ -1272,7 +1292,10 @@ int mbedtls_aes_crypt_cfb128( mbedtls_aes_context *ctx,
 
-    return( 0 );
+end:
+    if ( ret != 0 )
```
Style: Please remove space between `if` and `(`

andresag01 commented on library/aes.c:
```diff
@@ -1303,8 +1331,10 @@ int mbedtls_aes_crypt_cfb8( mbedtls_aes_context *ctx,
-
-    return( 0 );
+end:
+    if ( ret != 0 )
```
Style: Please remove space between `if` and `(`

andresag01 commented on library/aes.c:
```diff
@@ -1354,16 +1384,21 @@ int mbedtls_aes_crypt_ctr( mbedtls_aes_context *ctx,
-            mbedtls_aes_crypt_ecb( ctx, MBEDTLS_AES_ENCRYPT, nonce_counter, stream_block );
+            ret = mbedtls_aes_crypt_ecb( ctx, MBEDTLS_AES_ENCRYPT,
+                                         nonce_counter, stream_block );
+            if ( ret != 0 )
```
Style: Please remove space between `if` and `(`

andresag01 commented on library/aes.c:
```diff
@@ -1376,8 +1411,10 @@ int mbedtls_aes_crypt_ctr( mbedtls_aes_context *ctx,
-
-    return( 0 );
+end:
+    if ( ret != 0 )
```
Style: Please remove space between `if` and `(`

RonEld commented:
@andresag01 I believe I addressed your comments.
Please review

Patater commented on library/aes.c:
```diff
@@ -1287,11 +1310,16 @@ int mbedtls_aes_crypt_cfb8( mbedtls_aes_context *ctx,
     unsigned char ov[17];
+    int ret = 0;
+    const size_t orig_length = length;
+    unsigned char* out = output;
```
Use `unsigned char *out` instead of `unsigned char* out`.

Patater commented on library/aes.c:
```diff
@@ -1354,16 +1384,21 @@ int mbedtls_aes_crypt_ctr( mbedtls_aes_context *ctx,
+    int c, i, ret = 0;
     size_t n = *nc_off;
+    const size_t orig_length = length;
+    unsigned char* out = output;
```
Same here

RonEld commented:
@Patater I addressed your comments

sbutcher-arm commented:
@andresag01 - Can you please re-review this?

Patater commented on library/aes.c:
```diff
@@ -1376,8 +1411,10 @@ int mbedtls_aes_crypt_ctr( mbedtls_aes_context *ctx,
-    return( 0 );
+end:
+    if( ret != 0 )
+        mbedtls_platform_zeroize( out, orig_length );
```
Do we really need to zeroize `out`? Nothing has been written to `out` yet.

RonEld replied:
Same answer as https://github.com/ARMmbed/mbedtls/pull/1096/files#r222589243

Patater replied:
OK

Patater commented on library/aes.c:
```diff
@@ -1303,8 +1331,10 @@ int mbedtls_aes_crypt_cfb8( mbedtls_aes_context *ctx,
-    return( 0 );
+end:
+    if( ret != 0 )
+        mbedtls_platform_zeroize( out, orig_length );
```
Do we really need to zeroize `out`? Nothing has been written to `out` yet.

RonEld replied:
Same answer as https://github.com/ARMmbed/mbedtls/pull/1096/files#r222589243

Patater replied:
OK

Patater commented on library/aes.c:
```diff
@@ -1272,7 +1292,10 @@ int mbedtls_aes_crypt_cfb128( mbedtls_aes_context *ctx,
-    return( 0 );
+end:
+    if( ret != 0 )
+        mbedtls_platform_zeroize( out, orig_length );
```
Do we really need to zeroize `out`? Nothing has been written to `out` yet.

RonEld replied:
`out` is a pointer to the original `output`.
If, for example, In lines 1270 and 1287 we write to `output`, and by that updating `out`.
If for example, second iteration in the operation fails, we need to clear the whole output buffer. Or at least, what was written.

Patater replied:
OK

Yes, missed there was a loop here. All good.

RonEld replied:
Perhaps an optimization would be to clear only the number of bytes written, and not hte whole buffer. For example, if a 4GB buffer is given and the failure happens after 32 bytes, I don't see the point in clearing the full 4GB.

Patater commented on library/aes.c:
```diff
@@ -1065,7 +1072,10 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
-    return( 0 );
+end:
+    if( ret != 0 )
+        mbedtls_platform_zeroize( out, orig_length );
```
If we can count on `mbedtls_aes_crypt_ecb()` to zeroize `output` when it fails, I don't think we need to zeroize `out` again here, since it should already be zeroized.

RonEld replied:
This is in case we can count on that.
What if someone replaces only the `mbedtls_internal_aes_encrypt()` and `mbedtls_internal_aes_decrypt()` functions? For example, by using the Arm crypto extensions, and the operations fails in the middle? Can we assure that it clears the `out` buffer?

Patater replied:
OK

This is also in a loop which I didn't noticed before, so it makes sense to zeroize here.

Patater commented on library/ctr_drbg.c:
```diff
@@ -447,7 +458,7 @@ int mbedtls_ctr_drbg_random_with_add( void *p_rng,
 
     ctx->reseed_counter++;
```
Are you sure you want to increment the reseed counter when `ret != 0`? I'm assuming `ret` can be non-zero here, since you wrote `return( ret );`.

`return( 0 )` made it obvious that reseed_counter was only incremented on success, because control flow would only reach here on success. Now I have to read the function more carefully to see if any error cases fall through to here.

RonEld commented:
@Patater I optimized the code according to https://github.com/ARMmbed/mbedtls/pull/1096#discussion_r222622800

I also changed the return value to 0 in `ctr_drbg` to ake it clearer.

In addition, I modified the xts code as it was introduced after the original PR was written. Please look carefully at the xts code, as I am not entirely sure about it, specifically in the `leftover` part,

RonEld commented on library/aes.c:
```diff
@@ -1069,12 +1070,13 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
 end:
     if( ret != 0 )
-        mbedtls_platform_zeroize( out, orig_length );
+        mbedtls_platform_zeroize( out, bytes_written );
```
Personally, I prefer not to have variables, if not needed, but in this specific case, I think it's more readable this way.

sbutcher-arm replied:
To keep the code footprint down, we should really follow @Patater's suggestion.

RonEld commented:
@Patater I addressed your comments

sbutcher-arm commented:
@RonEld - I'm afraid the PR needs rework. Could you please rebase it?

RonEld commented:
@sbutcher-arm I think a rebase should be done after the PR was approved, to avoid confusion. Shouldn't it?

RonEld commented:
I have rebased to resolve conflicts.
cc @Patater

RonEld commented:
@Patater Thanks for the review!
I'll squash the commits once a second reviewer approves

sbutcher-arm commented on library/aes.c:
```diff
@@ -1045,8 +1045,10 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
                     const unsigned char *input,
```
Do we want to update the header documentation to indicate errors are propagated upwards from the ECB operation?

RonEld replied:
I don't think it's needed.
1. Our ECB operation returns success, so no error.
2. An alternative implementation will most likely not even call the ECB operation, and call directly the HW crypto engine

sbutcher-arm commented on library/aes.c:
```diff
@@ -1179,6 +1191,8 @@ int mbedtls_aes_crypt_xts( mbedtls_aes_xts_context *ctx,
     unsigned char tweak[16];
```
Again, do we want to update the header documentation to indicate errors are propagated upwards from the ECB operation?

RonEld replied:
same answer

sbutcher-arm commented on library/aes.c:
```diff
@@ -1075,7 +1077,9 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
         {
             memcpy( temp, input, 16 );
-            mbedtls_aes_crypt_ecb( ctx, mode, input, output );
+            ret = mbedtls_aes_crypt_ecb( ctx, mode, input, output );
```
I think it would be good to have some tests to inject errors into the function.

RonEld replied:
I agree, but `mbedtls_aes_crypt_ecb()` always return 0

sbutcher-arm replied:
I was actually imagining we replace `mbedtls_aes_crypt_ecb()` with a test equivalent to inject errors, but that's a bit ambitious and more of a test feature than a fix, and really outside the scope of this PR.

I can accept it as it is.

sbutcher-arm commented on library/aes.c:
```diff
@@ -1441,27 +1493,33 @@ int mbedtls_aes_crypt_ctr( mbedtls_aes_context *ctx,
 
     while( length-- )
     {
         if( n == 0 ) {
```
Can you please fix the bracing on the preceding line?

RonEld commented:
@sbutcher-arm Thanks for the review. 
I addressed your comments.

cc @Patater 

I will squash the commits once the PR is fully approved

sbutcher-arm commented:
I guess the PR just needs squashing, back porting and re-approval now.

RonEld commented:
@sbutcher-arm yesm that's my plan

@andresag01 \ @Patater  Do you want me to squash before you approve?

RonEld commented:
I removed the `bytes_written` in all other modes, except xts, as in this mode, `output` does not point to the end of the buffer, when the operation finishes.

RonEld commented:
Backports available at #2396 and #2397

mpg commented:
Almost ready for merge, just waiting for the 2.7 backport to be approved.

RonEld commented:
Both backports have been fully reviewed and approved, so removing the "needs backports" label"

cc @Patater

gilles-peskine-arm commented:
Unfortunately, there are now merge conflicts. I think I independently rediscovered the issue in ctr_drbg. The aes part is still relevant.

gilles-peskine-arm commented:
* Superseded by https://github.com/ARMmbed/mbedtls/pull/4760 for the AES result checks.
* Tracked internally for the missing zeroizations.
* Already done independently for the CTR_DRBG fixes.
